### DISCUSSION TITLE
Fullrow editing and accept/discard buttons

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/slickgrid-universal#1290] [Ideas]
hey there, AgGrid has a quite interesting feature called Fullrow Editing, see this example https://blog.ag-grid.com/full-row-editing-ag-grid-committing-changes-button-click/

I wonder if we can do something like that with slickgrid universal?

**Comment by ghiscoding** [comment]
Not possible in SlickGrid since only a single Editor can be open at a single time in the grid, that's why I created the optional Composite Editor feature

**Reply by zewa666** [reply]
so thats like a hard limitation from the core slickgrid dependency itself?

**Reply by ghiscoding** [reply]
yeah something like that, it was built that way I think, so it's kind hard to go around it because in the core lib there are many references to the active editor (single). So I assume it would require substantial changes to accept multiple cell being edited at the same time, it's all related to the active cell & editor and it's all in a single manner. With the latest major, I do have the core lib now in Slickgrid-Universal but I want to stay as close as possible to the official one (6pac/slickgrid), but even then I think this would require too many changes

**Reply by zewa666** [reply]
yep that sounds understandable.

If we leave out the multiple editors tough, adding custom buttons for edit, delete, save, cancel shouldnt be too hard. additionally the edited row could be hilighted to get a better indicator. now if I store the edited row on edit click, if there were a beforeEditorOpen callback sorta where I could conditionally cancel the editor popping up if its not part of the edited row we'd be in a similar place.

I could additionally check all editCommands made on the row and perhaps merge them on update click.

**Reply by ghiscoding** [reply]
The composite editor has a workaround to bypass these kind of problems, there's a target type that I added on the `onBeforeEditCell`, it can be `target: 'grid'` or `target: 'composite'`. Each editor also has extra checks when `compositeEditorOptions` are provided to the editor, for example below.... but again that is meant to be used in a modal window, not within the grid. I don't really have time to spend on this though

https://github.com/ghiscoding/slickgrid-universal/blob/3a8795c9827f7cc8cdacf5140fa198471f9d76a0/packages/common/src/editors/dateEditor.ts#L255-L281

**Reply by zewa666** [reply]
onBeforeEditCell was the missing link. I guess from there I have everything I need to build that functionallity. I really dont care whether all cells of the row are displayed as editors since you can edit anyways only cell by cell. btw I didnt mean you should implement this by yourself, I was merely searching for ideas how to approach this by myself

I guess a proper way to implement this in a reusable way would be to create a new plugin. do you have any recommendation which one to look at as blueprint? the custom tooltip one pooked quite promising

**Reply by ghiscoding** [reply]
I guess it depends how much of a plugin you want this to be? I mean, there are a few plugins that I've put directly in the `packages/common` because they are too common to be optional (e.g. HeaderMenu, ContextMenu) but they really are optional in SlickGrid itself. If it's small then I guess you can put it in the common package but if it's big then optional packages are ExcelExport, TextExport, RowDetail, OData and GraphQL), the TextExport should be relatively easy to follow and if you go that route then you'll have to use `externalResources` from grid option.  ahhh yes the Custom Tooltip is also another small one

**Reply by zewa666** [reply]
great examples, yeah I think I'll start with the common route and see where it ends. if it becomes too large I can always move it into something custom. thx for the guidance so far

**Reply by ghiscoding** [reply]
if you go with the common package, then all plugins were moved into `common/extensions`, I named it extensions because in SlickGrid they have Controls (columnPicker, gridMenu) and Plugins (contextMenu, headerMenu, ...) so extension was a name that encapsulate both. After you create a new class in the extension folder, you will then go with the Extension Service which is where they get instantiated, I typically went with a flag name (e.g. `enableContextMenu`) and when that flag is found, then the ExtensionService does the rest by instantiated the plugin and keep ref of it

**Reply by zewa666** [reply]
a quick question about examples, should I create a new one or add the plugin to an existing?

edit: oh and I didnt yet find a way how to access the gridService from within the plugin. neither was the containerservice available. any ideas how to get that?

**Reply by ghiscoding** [reply]
you can reuse an example if you want and if it's not already too heavy on the Cypress tests side. 

Why do you want to access the GridService for? The plugins are typically kinda of standalone and typically only use the grid/dataview objects (through `init()`), I created the GridService as a helper when I was tired of repeating the same code for CRUD stuff and then expended it over the years. So you might remember that it's not a true DI Container, all the services passed to the constructor are passed in the `slick-vanilla-grid-bundle.ts` where each of them are being instantiated but a very important note is that because it's not a true DI, the order they are instantiated really matter (same goes for Angular-Slickgrid). As for the ContainerService it's only used for optional package, it's never used within the common package.  

https://github.com/ghiscoding/slickgrid-universal/blob/21d1489df6afaa07b7950d0c614c143f904cb9d7/packages/vanilla-bundle/src/components/slick-vanilla-grid-bundle.ts#L345-L357

**Reply by zewa666** [reply]
well exactly due to crud support for deleting a row since the new actions column added would contain an edit and delete button for each row. but you're right I might be better of just looking up the delete methods implementation.

yeah I noticed its no true DI container reading the comments.

**Reply by zewa666** [reply]
hmm actually it felt too wrong to re-implement all the details that are to be found in the delete method so I've sticked to the gridService and routing it through the bundle over to the extensionService and to the init of my plugin. This one could certainly need some improvement.

Nevertheless, I wanted to show you what I got so far after a quick stab https://github.com/zewa666/slickgrid-universal/commit/d4372aae30e774bff4d6ad577779ae93d79fdba2 which is not done but a first look so that you get a chance to say whether this feels too specific, in which case I'd just implement it locally for our apps, or it could be a generic feature.

**Reply by ghiscoding** [reply]
Funny enough, I was actually going to ping you and ask how it is going with this because I wanted to possibly push a new release tomorrow to close up all opened PRs and push a few fixes. I can wait a few more days though since you seem to be making good progress. Side note, I've added an ESLint rule to prefer single quote, so things like this `pluginName = "RowBasedEdit" as const;` will need to change. 

without actually doing a git checkout and running the code yet, would it be possible to do a print screen or animated gif of that feature?

**Reply by zewa666** [reply]
I dont think I'll make it on time. throughout the implementation of the POC I've realized there might be tons of parts the user might want to override. like you mentioned the position of the col, the template of the buttons etc.

So I'd like to take on this a bit slower and think through cases.

Also I'm not next to the PC but can post a gif tomorrow.

**Comment by zewa666** [comment]
Ok so here's the quick preview I have for the plugin

![row-based-edit](https://github.com/ghiscoding/slickgrid-universal/assets/2861414/b80321e3-08a7-4583-8bd9-3f68244fc2ad)

**Reply by ghiscoding** [reply]
oh ok, I thought you had found a way to make the entire row editable at once but it's not entirely that. I think that in that case it might be good if we could have a way to add some CSS classes to all the cells that are allowed to be editable. With the CSS classes, the user could at least know which one are editable (clickable). So I'm not sure if it's actually doable, it's just a suggestion on the UX side of things, if doable then these classes could be customized by the user (or just simple CSS/SASS variables might suffice)

This also kind of remind me of batch editing, which is what I've done in Example 12, all the cells that got edited (even to empty) are turning with an orange outline and once they're all saved then I remove that outline (via ItemMetadata styling)

**Reply by zewa666** [reply]
agreed, thats what I started doing with adding a `slick-rbe-editmode` class to rows in editmode and updating the themes to define two new css vars and defaulting different background for those rows. Also I've switched from rows to id/datasetIdPropertyName for the indicator vs the row index.

also as mentioned initially, the difference is that now you're only allowed to edit cells within rows in editmode (either one or multiple, depending on the config)

 https://github.com/zewa666/slickgrid-universal/commit/8adc3dd38b41033057c12685961fcaa6576c951c

So in example 11 (batch updates) the usage of setCellCssStyles is another missing piece I was looking for. I went with monkey patching `getItemMetadata`  instead, which didn't work though as thought since it's only triggered on rerendering rows. I thought instead of doing so I'd manipulate the DOM directly (you can see all of that in the above linked commit) to spare the extra re-render roundtrip

**Reply by ghiscoding** [reply]
pretty damn cool, I'm always impressed by new approach like this that I didn't think about :)  SlickGrid is so customizable, that is also why I love the lib so much. Looking good so far 👍🏻 

It looks like you're similar to me, I'm on holidays since a few days but I still work on the lib lol

**Reply by zewa666** [reply]
well ... guilty ;)

holiday with two kids is mildly said challenging, happy to have the nights blocked for coding fun 😅

so if you're still fine with turning this into an official plugin I think next I'll create a github issue to track tasks and progress.

Also with regards to the examples I didnt really find any suitable (either wrong context or too specific cypress tests) so if you're fine with it I'd add a new one. Additionally, could you point me to where in the docs you would describe this plugin? I thought about adding a new page in the grid functionallities section

**Comment by ghiscoding** [comment]
> so if you're still fine with turning this into an official plugin I think next I'll create a github issue to track tasks and progress.

it seems kinda small though, up to you if you want to create a new plugin (package) but I'd be ok also to keep it as an internal extension that only requires a flag to enable (an external package will require `externalResource` to be provided)

Adding a new Example seems reasonable, I'm not sure when you forked the repo but I added Example 21 couple days ago, so that would require you to use Example 22. 

> could you point me to where in the docs you would describe this plugin? I thought about adding a new page in the grid functionallities section

yeah I guess Grid Functionalities seems like a good place. We could also perhaps add a reference into the main Editors.md doc

**Reply by zewa666** [reply]
sry if I was unclear. It wasn't about whether it's part of common or an additional new package. I wasn't sure if you'd like this feature to be a plugin at all, such as whether it can be of value to others as well.

**Reply by ghiscoding** [reply]
yes I definitely think this can be of value to others, so yes it could be part of the common package I think

**Comment by zewa666** [comment]
closing this in favor of https://github.com/ghiscoding/slickgrid-universal/issues/1299
